pkg/vminfo/syscalls.go

```diff
@@ -0,0 +1,372 @@
+type checkContext struct {
+	impl    checker
+	cfg     *mgrconfig.Config
+	target  *prog.Target
+	sandbox ipc.EnvFlags
+	// Checkers use requests channel to submit their test programs,
+	// main goroutine will wait for exactly pendingRequests message on this channel
+	// (similar to sync.WaitGroup, pendingRequests is incremented before starting
+	// a goroutine that will send on requests.
+	requests        chan []*rpctype.ExecutionRequest
+	pendingRequests int
+	// All collected test programs that will be sent to VMs.
+	progs []rpctype.ExecutionRequest
+	// Ready channel is closed after we've recevied results of execution of test
+	// programs and file contents. After this results maps and fs are populated.
+	ready   chan bool
+	results map[int64]*ipc.ProgInfo
+	fs      filesystem
+	// Once checking of a syscall is finished, the result is sent to syscalls.
+	// The main goroutine will wait for exactly pendingSyscalls messages.
+	syscalls        chan syscallResult
+	pendingSyscalls int
+}
+
+type syscallResult struct {
+	call   *prog.Syscall
+	reason string
+}
+
+func newCheckContext(cfg *mgrconfig.Config, impl checker) *checkContext {
+	sandbox, err := ipc.SandboxToFlags(cfg.Sandbox)
+	if err != nil {
+		panic(fmt.Sprintf("failed to parse sandbox: %v", err))
+	}
+	ctx := &checkContext{
+		impl:     impl,
+		cfg:      cfg,
+		target:   cfg.Target,
+		sandbox:  sandbox,
+		requests: make(chan []*rpctype.ExecutionRequest),
+		results:  make(map[int64]*ipc.ProgInfo),
+		syscalls: make(chan syscallResult),
+		ready:    make(chan bool),
+	}
+	ctx.startCheck()
+	return ctx
+}
+
+func (ctx *checkContext) startCheck() {
+	for _, id := range ctx.cfg.Syscalls {
+		call := ctx.target.Syscalls[id]
+		if call.Attrs.Disabled {
+			continue
+		}
+		ctx.pendingSyscalls++
+		syscallCheck := ctx.impl.syscallCheck
+		if strings.HasPrefix(call.CallName, "syz_ext_") {
+			// Non-mainline pseudo-syscalls in executor/common_ext.h can't have
+			// the checking function and are assumed to be unconditionally supported.
+			syscallCheck = alwaysSupported
+		}
+		// HostFuzzer targets can't run Go binaries on the targets,
+		// so we actually run on the host on another OS. The same for targets.TestOS OS.
+		if ctx.cfg.SysTarget.HostFuzzer || ctx.target.OS == targets.TestOS {
+			syscallCheck = alwaysSupported
+		}
+
+		var depsReason chan string
+		deps := ctx.cfg.SysTarget.PseudoSyscallDeps[call.CallName]
+		if len(deps) != 0 {
+			ctx.pendingRequests++
+			depsReason = make(chan string, 1)
+			go func() {
+				depsReason <- ctx.supportedSyscalls(deps)
+			}()
+		}
+		ctx.pendingRequests++
+		go func() {
+			reason := syscallCheck(ctx, call)
+			ctx.waitForResults()
+			if reason == "" && depsReason != nil {
+				reason = <-depsReason
+			}
+			ctx.syscalls <- syscallResult{call, reason}
+		}()
+	}
+	dedup := make(map[hash.Sig]int64)
+	for i := 0; i < ctx.pendingRequests; i++ {
+		for _, req := range <-ctx.requests {
+			sig := hashReq(req)
+			req.ID = dedup[sig]
```

Comment:
Is there a high chance of duplicate progs?

Reply:
Currently there are 4349 test programs for Linux + 105 duplicates.
Initially I expected there will be more b/c each check is independent and just asks to run some program w/o looking what other checkers are doing. So if lots of checkers use the same helper, say "if usb injection available", they will all issue the same test program.